### PR TITLE
feat: add QUIC transport to the chat example

### PR DESCRIPTION
**Pull request by yellowred**
## Description

See: https://github.com/libp2p/rust-libp2p/issues/3501#issuecomment-1461492407

## Notes & open questions
Removed `development_transport` and used a custom transport, since we don't want Quic in dev transport.
However, I anticipate others will copy this code to their new projects when they need Quic+Tcp, so there is a question of the purpose of the `development_transport` function.

## Change checklist

- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] A changelog entry has been made in the appropriate crates

### CODE REVIEW

File: examples/chat-example/src/main.rs [hunk 1]
```diff
@@ -46,34 +46,49 @@
...
-    struct MyBehaviour {
-        gossipsub: gossipsub::Behaviour,
-        mdns: mdns::async_io::Behaviour,
-    }
+    let tcp_transport = tcp::async_io::Transport::new(tcp::Config::default().nodelay(true))
+        .upgrade(upgrade::Version::V1)
+        .authenticate(
+            noise::NoiseAuthenticated::xx(&id_keys).expect("signing libp2p-noise static keypair"),
+        )
+        .multiplex(mplex::MplexConfig::new())
```
Comment (thomaseizinger) [inline examples/chat-example/src/main.rs]: `yamux` is the preferred muxer on top of TCP so let's use that please.

File: examples/chat-example/src/main.rs [hunk 2]
```diff
@@ -46,34 +46,49 @@
...
+        .timeout(std::time::Duration::from_secs(20))
+        .boxed();
+    let quic_transport = quic::async_std::Transport::new(quic::Config::new(&id_keys));
+    let transport = OrTransport::new(quic_transport, tcp_transport)
+        .map(|either_output, _| match either_output {
+            Either::Left((peer_id, muxer)) => (peer_id, StreamMuxerBox::new(muxer)),
+            Either::Right((peer_id, muxer)) => (peer_id, StreamMuxerBox::new(muxer)),
+        })
```
Comment (thomaseizinger) [inline examples/chat-example/src/main.rs:91]: Hopefully, this will get simpler eventually once we tackle https://github.com/libp2p/rust-libp2p/issues/3179.

File: examples/chat-example/src/main.rs [hunk 3]
```diff
@@ -114,7 +127,8 @@ async fn main() -> Result<(), Box<dyn Error>> {
     let mut stdin = io::BufReader::new(io::stdin()).lines().fuse();
 
     // Listen on all interfaces and whatever port the OS assigns
-    swarm.listen_on("/ip4/0.0.0.0/tcp/0".parse()?)?;
+    swarm.listen_on("/ip4/0.0.0.0/udp/0/quic-v1".parse()?)?;
+    // swarm.listen_on("/ip4/0.0.0.0/tcp/0".parse()?)?;
```
Comment (thomaseizinger) [inline examples/chat-example/src/main.rs]: I think doing both would be useful.

File: examples/chat-example/src/main.rs [hunk 4]
```diff
@@ -149,6 +163,9 @@ async fn main() -> Result<(), Box<dyn Error>> {
                         "Got message: '{}' with id: {id} from peer: {peer_id}",
                         String::from_utf8_lossy(&message.data),
                     ),
+                SwarmEvent::NewListenAddr { address, .. } => {
+                    println!("Local node is listening on {:?}", address);
```
Comment (thomaseizinger) [inline examples/chat-example/src/main.rs]:
```suggestion
                    println!("Local node is listening on {address}");
```

This will be nicer to look at.

File: examples/chat-example/src/main.rs [hunk 5]
```diff
@@ -46,34 +46,49 @@
...
 use std::error::Error;
 use std::hash::{Hash, Hasher};
 use std::time::Duration;
 
+// We create a custom network behaviour that combines Gossipsub and Mdns.
+#[derive(NetworkBehaviour)]
+struct MyBehaviour {
+    gossipsub: gossipsub::Behaviour,
+    mdns: mdns::async_io::Behaviour,
+}
```
Comment (thomaseizinger) [inline examples/chat-example/src/main.rs:68]: Why is this here if we don't use it?

**Review by thomaseizinger — COMMENTED** [review]
Thank you! A few questions.

File: examples/chat-example/src/main.rs (hunk 5 above)
Comment (yellowred) [reply examples/chat-example/src/main.rs:68]: `MyBehaviour` is used below, I just moved it from the `main`.

File: examples/chat-example/Cargo.toml [hunk 6]
```diff
@@ -1,6 +1,6 @@
 [package]
 name = "chat-example"
-version = "0.1.0"
+version = "0.2.0"
```
Comment (thomaseizinger) [inline examples/chat-example/Cargo.toml]: No need to bump this, we are not releasing them.

File: examples/chat-example/src/main.rs (hunk 5 above)
Comment (thomaseizinger) [reply examples/chat-example/src/main.rs:68]: Sorry, my bad.

**Review by thomaseizinger — COMMENTED** [review]
One request, other LGTM

**Comment by thomaseizinger** [comment]
> Removed development_transport and used a custom transport, since we don't want Quic in dev transport.\nHowever, I anticipate others will copy this code to their new projects when they need Quic+Tcp, so there is a question of the purpose of the development_transport function.

It is a legitimate question. Personally, I think users should compose their transport for a production application themselves anyway.

But, it is useful to get started quickly.

Maybe we can incorporate this into `SwarmBuilder` somehow?

An idea sketch would be:

```rust
SwarmBuilder::new(local_peer_id)
   .tokio_executor()
   .default_transport()
   .behaviour(my_behaviour)
   .build()
```

The builder would be type-state guided such that you must call either `default_transport` or `custom_transport(my_transport)`.

**Comment by yellowred** [comment]
> But, it is useful to get started quickly.
> 
> Maybe we can incorporate this into `SwarmBuilder` somehow?

I like this idea. Everyone will be able to start quickly with a default and customize later on when they see a need for that.

**Review by thomaseizinger — APPROVED** [review]

**Comment by thomaseizinger** [comment]
The clippy build is failing due to a recent PR that merged into master (https://github.com/libp2p/rust-libp2p/pull/3580), can you address that please?

**Comment by mxinden** [comment]
Idea above for a `default_transport` method on `Swarm` or an extension trait to `Swarm` sounds good to me. I don't have an opinion on the exact implementation of the mechanism. I do believe providing some default is valuable. Anecdotally just yesterday in the community call someone asked for the best defaults on a transport to get started.

**Comment by thomaseizinger** [comment]
> Idea above for a `default_transport` method on `Swarm` or an extension trait to `Swarm` sounds good to me. I don't have an opinion on the exact implementation of the mechanism. I do believe providing some default is valuable. Anecdotally just yesterday in the community call someone asked for the best defaults on a transport to get started.

As you are pointing out, it would have to be an extension trait because `libp2p-swarm` does not pull in any specific transports. I'll open an issue.

**Comment by yellowred** [comment]
@thomaseizinger Added Quic as a separate crate.

**Review by thomaseizinger — APPROVED** [review]
